python_modules/dagster/dagster/_core/definitions/unresolved_asset_sensor_definition.py

```diff
@@ -0,0 +1,279 @@
+
+        return cursor_update_dict
+
+    def _get_cursor(self, cursor, asset_key):
+        return cursor.get(str(asset_key)) if cursor.get(str(asset_key)) else {}
+
+    def _make_sensor(self, upstream: Mapping[AssetKey, Set[AssetKey]]) -> SensorDefinition:
+        """Creates the sensor that will monitor the parents of all provided assets and determine
+        which assets should be materialized (ie their parents have been updated).
+
+        The cursor for this sensor is a dictionary mapping stringified AssetKeys to dictionaries. The
+        dictionary for an asset is a mapping of the stringified AssetKeys of its parents to the storage id
+        of the latest materialization that has resulted in the child asset materializing
+        """
+
+        def sensor_fn(context):
+
+            cursor_dict = json.loads(context.cursor) if context.cursor else {}
+            should_materialize = []
+            cursor_update_dict = {}
+
+            # sort the assets topologically so that we process them in order
+            toposort_assets = list(toposort.toposort(upstream))
+            # unpack the list of sets into a list
+            toposort_assets = [
+                asset for layer in toposort_assets for asset in layer if asset in upstream.keys()
+            ]
+
+            # determine which assets should materialize based on the materialization status of their
+            # parents
+            for a in toposort_assets:
+                a_cursor = self._get_cursor(cursor_dict, a)
+                parent_materialization_records = self._get_parent_materializations(
+                    context,
+                    parent_assets=upstream[a],
+                    cursor=a_cursor,
+                    will_materialize_list=should_materialize,
+                )
+
+                if all(
+                    [
+                        materialization_status
+                        for materialization_status, _ in parent_materialization_records.values()
+                    ]
+                ):  # TODO should we allow the user to specify all() vs any() here?
+                    should_materialize.append(a)
+                    cursor_update_dict[str(a)] = self._make_cursor_update_dict(
+                        parent_event_records=parent_materialization_records, current_cursor=a_cursor
+                    )
+                else:
+                    # if the asset shouldn't be materialized, then keep the cursor the same
+                    cursor_update_dict[str(a)] = self._get_cursor(cursor_dict, a)
+
+            if len(should_materialize) > 0:
+                context.update_cursor(json.dumps(cursor_update_dict))
+                return RunRequest(run_key=f"{context.cursor}", asset_selection=should_materialize)
+
+        return SensorDefinition(
+            evaluation_fn=sensor_fn,
+            name=self.name,
+            job_name="__ASSET_JOB",
```

Comment:
take a look at the logic here: https://github.com/dagster-io/dagster/blob/master/python_modules/dagster/dagster/_core/definitions/asset_group.py#L433

I can add more detail if the logic there isn't clear

Reply:
oh interesting, so there's an `_ASSET_JOB` for each kind of partition (daily, yearly, etc)? 

thinking out loud here...
so we'll need a way to pass all of those jobs to the SensorDefinition, which will be complicated because we won't know how many partitioned jobs there are until the repository is loaded (and since we're likely moving away from the UnresolvedAssetSensor this means we won't know what jobs to pass to the sensor definition at construction time)